### PR TITLE
Support for python 3 in sascalc

### PR DESCRIPTION
**Pull request by pkienzle**
With this PR the tests now run in python 3.

This PR already merges in the branches for ticket-915, ticket-869, ticket-887 and ticket-853.  Either merge them separately in the above order or just apply this one and delete the other PRs.  Note that ticket-915 and ticket-853 do not yet have PRs.

## Details of the python 2-3 conversion

1. New syntax for raise in python 3:

   ```raise Exception, message  => raise Exception(message)```

   I also did some linting:

   ```except:  =>  except Exception:```

2. Relative imports require '.' in python 3:

     ```import relative => import .relative```

    I also rearranged imports according to pep8 general -> specific rules, with builtins first, third party second, sister directories third and imports from the same directory last.

3. Convert absolute imports into relative imports:

   ```from sas.sascalc.dataloader.module  => from ..module```

    Not necessary.  Just makes tree rearrrangement a little easier.

4. bytes vs. unicode

    I added some  methods if filer reader base class to read the file and return strings, with python 2 returning the file return values directly, and python 3 decoding the bytes as strings:

    ```
    self.f_open.read()  => self.readll()
    self.f_open.readline() => self.nextline()
    self.f_open.readlines() => self.nextlines()
    ```

    Similarly, h5 attributes and string values needed to be converted, as well as xml etree.tostring() results.

5. list.sort() has new interface

    ```list.sort(lambda a, b: key(a) < key(b))  =>  list.sort(key=key)```

6. assume nxsunit is available

    I don't understand why nxsunit is an "optional" import, with code like:

    ```
    has_converter = False
    try:
        from sas.sacalc.data_util.nxsunit import Converter
    except:
        has_converter = false
    ...
    if has_converter and unit != target_unit:
        ...
   ```

    becomes

    ```
    from sas.sascalc.data_util.nxsunit import Converter
    ...
    if unit != target_unit:
        ...
    ```

    This was not a necessary change, just code cleanup.  I did not update all readers.

7. dict.iteritems no longer exists

    ```for k, v in dict.iteritems():  =>  for k, v in dict.items():```

    On rare occasions when dict is changing while processing the items, you need to use ```list(dict.items())``` instead.

8. exec is no longer a statement

    Exec should generally be avoided, so for the cansas reader I converted it into a call to getattr/setattr, with checks for dots in the attr name, using functions getattrchain/setattrchain that I wrote:

    ```
    exec "localvar = container.%s" % variable => localvar = getattrchain(container, variable)
    exec "container.%s = %g" % (variable, value) => setattrchain(container, variable, value)
    ```

    I also added the following TODO:

    ```
    # TODO: why split() when accessing unit, but not when setting value?
    ```

    since the the unit was retrieved from variable.split('.')[0]+"_unit" but the value was being set to variable without the split.

    The other use of exec was in readers.associations.  For that I just did:

    ```exec "statements" => exec(statements)```

    It would be better to use importlib for this code.

9. xrange is no longer

    ```',',join([str(vector[i]) for in in xrange(len(vector))]) => ','.join(str(v) for v in vector)```

10. map returns an iterator instead of a list

    ```converted = map(function, list) => converted = list(map(function, list))```

    You don't need ```list(iter)``` if you can use the iter directly, but apparently numpy chose to store the iterator as an object rather than an array of iterated values.  This is a surprise; it should be testing if it is a sequence and expanding the sequence.

11. clean up latex expansion

    ```"\\c" => r"\c"```

    Python no longer accepts \c when c is not one of the standard backslash coded characters such are ```\r``` and ```\n```.

12. C api updates

    C APIs have changed a bit, but with a couple of macros and a tiny bit of repeated code both 2 and 3 can be supported.  I didn't use the new frozen ABI which should allow compiled modules to run under any version of python 3.  Probably cannot do so with our use of numpy.

    Still one warning left (when including numpy arrays), so can't close ticket 4 :-(

13. thread is no longer available

    ```import thread => import _thread as thread```

### CODE REVIEW

**Comment by pkienzle** [comment]
ready for testing on OSX

**Comment by pkienzle** [comment]
ready for testing on Win32

**Comment by krzywon** [comment]
For note keeping, the following tickets are addressed with this PR: [14](http://trac.sasview.org/ticket/14), [853](http://trac.sasview.org/ticket/853), [869](http://trac.sasview.org/ticket/869), [887](http://trac.sasview.org/ticket/887), [903](http://trac.sasview.org/ticket/903), [915](http://trac.sasview.org/ticket/915), [985](http://trac.sasview.org/ticket/985), [994](http://trac.sasview.org/ticket/994), and [997](http://trac.sasview.org/ticket/997).

**Comment by krzywon** [comment]
When trying to save the report results window as a PDF, 2 PNG files were saved, one of the fit panel and another of the residuals panel. No PDF file with the parameter values was saved.

File: src/sas/sascalc/dataloader/readers/abs_reader.py [hunk 1]
```diff
@@ -108,9 +103,14 @@ def get_file_contents(self):
 
                 # Sample thickness in mm
                 try:
-                    value = float(line_toks[5])
-                    if self.has_converter and \
-                            self.current_datainfo.sample.thickness_unit != 'cm':
+                    # ABS writer adds 'C' with no space to the end of the
+                    # thickness column.  Remove it if it is there before
+                    # converting the thickness.
+                    if line_toks[5][:-1] not in '012345679.':
```
[the remark this answers is not in the document]
Comment (pkienzle) [reply src/sas/sascalc/dataloader/readers/abs_reader.py]: Yes of course!  Isn't it lovely how the writers eat all the errors so we can't see them?!

File: src/sas/sascalc/dataloader/readers/cansas_reader.py [hunk 2]
```diff
@@ -1288,37 +1236,26 @@ def _store_float(self, location, node, variable, storage, optional=True):
             units = entry.get('unit')
             if units is not None:
                 toks = variable.split('.')
-                local_unit = None
-                exec "local_unit = storage.%s_unit" % toks[0]
+                # TODO: why split() when accessing unit, but not when setting value?
```
[the remark this answers is not in the document]
Comment (pkienzle) [reply src/sas/sascalc/dataloader/readers/cansas_reader.py]: Please replace the TODO with a comment explaining that the unit names are stored at the top level of storage even if the values are stored in a subobject.

**Comment by pkienzle** [comment]
ready for testing on Win32

**Comment by pkienzle** [comment]
ready for testing on Win32

**Comment by pkienzle** [comment]
I modified the behaviour of the saved report a bit:

* for .txt format I save images as .png rather than .pdf

* I use the same image name ```basename_img#.png``` for all of .txt, .html and .pdf; each format used different image names before.

* The .pdf format deletes the images after the pdf is created, so if you save as .html then as .pdf with the same base name then the .html will be broken.

* could use a data uri so the intermediate files are not needed for html or pdf, which would solve the above problem:

```
data_uri = open('11.png', 'rb').read().encode('base64').replace('\n', '')
img_tag = '<img src="data:image/png;base64,{0}">'.format(data_uri)
```

**Comment by rozyczko** [comment]
ready for testing on Win32

**Comment by rozyczko** [comment]
ready for testing on Win64

**Comment by wpotrzebowski** [comment]
ready for testing on Win64

**Comment by wpotrzebowski** [comment]
ready for testing on Win64

**Comment by wpotrzebowski** [comment]
ready for testing on Win64

**Comment by rozyczko** [comment]
ready for testing on Win64

**Comment by wpotrzebowski** [comment]
ready for testing on Win64

**Comment by wpotrzebowski** [comment]
ready for testing on Win64

**Comment by wpotrzebowski** [comment]
ready for testing on Win64

**Comment by krzywon** [comment]
I downloaded the Win64 installer and tested it today. The Invariant perspective is the only perspective that is available when running SasView. Fitting, P(r), and Corfunc are missing. I haven't investigated the issue yet, but this should be fixed before merging.

**Comment by wpotrzebowski** [comment]
ready for testing on Win64

**Comment by krzywon** [comment]
The newest Win64 installer still only loads the Invariant perspective. No errors are thrown in the sasview log file, but many of the information lines are missing, suggesting some part of sasview isn't being run. I attached the log as a reference.

The top half of the log was generated when running the Win64 build in this PR. The bottom half was generated when running release v4.1.2.
[sasview.log](https://github.com/SasView/sasview/files/1391276/sasview.log)

**Comment by rozyczko** [comment]
ready for testing on Win64

**Comment by rozyczko** [comment]
ready for testing on Win64

**Comment by rozyczko** [comment]
ready for testing on Win64

**Comment by krzywon** [comment]
ready for testing on Win32

**Comment by krzywon** [comment]
The latest Win32 build installs SasView correctly and executes with no obvious problems. The latest Win64 build still installs everything, but SasView.exe either does nothing or immediately closes itself. I suggest we merge this PR (after merging #110 into here) and work on the Win64 build issues separately.

**Comment by rozyczko** [comment]
ready for testing on Win64

**Comment by rozyczko** [comment]
ready for testing on Win64

**Comment by rozyczko** [comment]
ready for testing on Win64

**Comment by rozyczko** [comment]
ready for testing on Win64

**Pull request merged**